awscli/customizations/s3/factory.py

```diff
@@ -69,11 +101,42 @@ def create_transfer_manager(self, params, runtime_config,
 
     def _compute_transfer_client_type(self, params, runtime_config):
         if params.get('paths_type') == 's3s3':
-            return constants.DEFAULT_TRANSFER_CLIENT
-        return runtime_config.get(
-            'preferred_transfer_client', constants.DEFAULT_TRANSFER_CLIENT)
+            return constants.CLASSIC_TRANSFER_CLIENT
+        preferred_transfer_client = runtime_config.get(
+            'preferred_transfer_client',
+            constants.AUTO_RESOLVE_TRANSFER_CLIENT
+        )
+        if preferred_transfer_client == constants.AUTO_RESOLVE_TRANSFER_CLIENT:
+            return self._resolve_transfer_client_type_for_system()
+        return preferred_transfer_client
+
+    def _resolve_transfer_client_type_for_system(self):
+        transfer_client_type = constants.CLASSIC_TRANSFER_CLIENT
+        is_optimized_for_system = awscrt.s3.is_optimized_for_system()
+        LOGGER.debug(
+            'S3 CRT client optimized for system: %s', is_optimized_for_system
+        )
+        if is_optimized_for_system:
+            is_running = self._is_crt_client_running_in_other_aws_cli_process()
+            LOGGER.debug(
+                'S3 CRT client running in different AWS CLI process: %s',
+                is_running
+            )
+            if not is_running:
+                transfer_client_type = constants.CRT_TRANSFER_CLIENT
+        LOGGER.debug(
+            'Auto resolved s3 transfer client to: %s', transfer_client_type
+        )
+        return transfer_client_type
+
+    def _is_crt_client_running_in_other_aws_cli_process(self):
+        # If None is returned from acquiring the CRT process lock, it
+        # means the CRT S3 client is currently being used in a different
+        # AWS CLI process.
+        return acquire_crt_s3_process_lock() is None
 
     def _create_crt_transfer_manager(self, params, runtime_config):
+        acquire_crt_s3_process_lock()
```

Comment:
oh, clever acquiring the lock here AND acquiring the lock in the `_resolve_transfer_client_type_for_system()`

`_resolve_transfer_client_type_for_system()` will only run if the CLI is doing transfer-manager stuff, right? Another CLI instance doing something non-s3-related isn't going to hold the CrossProcessLock, is it?

Reply:
Yeah it should be fine. Specifically, this code path can only ever potentially be hit if we a `s3` transfer command (e.g. ``s3 cp``, ``s3 sync``, ``s3 mv``, and ``s3 rm``) is entered at the command line.